engine/source_sym_engine.ml

```diff
@@ -0,0 +1,184 @@
+  in
+  let bprint_source_expr buf = function SBlackbox s -> bprintf buf "%s" s in
+  let rec bprint_list ~sep bprint buf = function
+    | [] -> ()
+    | [x] -> bprint buf x
+    | x :: xs ->
+       bprintf buf "%a%t%a"
+         bprint x
+         sep
+         (bprint_list ~sep bprint) xs in
+  let break ntabs buf =
+    bprintf buf "\n%s" (BatList.init ntabs (fun _ -> "\t") |> String.concat "") in
+  let rec bprint_tree ntabs buf tree =
+    let sep = break (ntabs+1) in
+    match tree with
+    | Leaf (pilist, expr) ->
+      bprintf buf
+        "Leaf='%a':\
+         %t"
+        bprint_source_expr expr
+        (break (ntabs+1)) ;
+        bprintf buf "constraints {%a}" (bprint_list ~sep bprint_pi) pilist
+    | Node (pi, cst_list) ->
+      bprintf buf
+         "Node: constraints {%a}\
+          %t%a"
+         bprint_pi pi
+         (break (ntabs+1))
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+    | Conjunction (cst_list) ->
+      bprintf buf
+         "Conjunction:%t%a"
+         (break (ntabs+1))
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+    | Root cst_list ->
+       bprintf buf
+         "Root=\
+          %t%a"
+         (break (ntabs+1))
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+  in
+  let buf = Buffer.create 42 in
+  bprint_tree 0 buf stree;
+  BatIO.write_line BatIO.stdout (Buffer.contents buf)
+
+let sym_exec (source: source_program) =
+  let shares_kst ?wildcard_to_none:(w=false) k (clause:clause): clause option =
+    let pattern, expr = clause in 
+    match pattern with
+    | Wildcard when w = false-> Some (Wildcard, expr)
+    | Constructor (k', plist) when k' = k ->
+      if (List.length plist > 1) then
+        Some (Constructor (Tuple, plist), expr)
+      else
+        Some (plist |> List.hd, expr)
+    | _ -> None
+  in
+  let negate_piop = function
+    | EqInt i -> NeqInt i
+    | NeqInt i -> EqInt i
+    | IsK s -> NotK s
+    | NotK s -> IsK s
+    | IsTrue -> IsFalse
+    | IsFalse -> IsTrue
+    | IsTuple -> NotTuple
+    | NotTuple -> IsTuple
+    | IsNil -> NotNil
+    | NotNil -> IsNil
+    | IsCons -> NotCons
+    | NotCons -> IsCons
+    | EqString s -> NeqString s
+    | NeqString s -> EqString s
+  in
+  let piop_of_constructor = function
+    | Variant s -> IsK s
+    | Int i -> EqInt i
+    | Bool b -> if b then IsTrue else IsFalse
+    | String s -> EqString s
+    | Tuple -> IsTuple
+    | Nil -> IsNil
+    | Cons -> IsCons
+  in
+  let rec eval_clauses ?idx:(idx=0) constraints clauses : constraint_tree list =
+    match clauses with
+    | clause::tl -> let pattern, exp = clause
+      in
+      begin match pattern with
+        | Or (p1, p2) ->
+          eval_clauses constraints ((p1,exp)::(p2,exp)::tl) 
+        | Constructor (k, plist) ->
+          let pi = {idx=idx; op=piop_of_constructor k} in
+          let notpi = {pi with op=negate_piop pi.op} in 
+          let constraints' = (notpi :: constraints) in
```

Comment:
I'm reading the code as I go, so I don't have the full mental picture yet, but: if we can store constraints in the nodes of the tree, why do we need to collect all the constraints along the current path? (This sounds useful in representations where there is no tree structure, just a list of "leaves" with a constraint set and a right-hand-side.)